### ISSUE TITLE
`vmap` slower than manual vectorization part 2

### ISSUE THREAD

**Issue by AnkitSiva** [google/jax#7449]
Hey folks, I'm seeing an interesting issue that's a follow up to [this closed issue]( https://github.com/google/jax/issues/6312).

I was trying out the auto vectorization to move some of our hand batched code. The intent of the code block below is to multiply k-size (here 3) subsets of the n columns of a dataset and sum them to see what proportion of those datasets are all ones in those k columns. 
```python
from jax import numpy as np, vmap, jit, random
import itertools

key = random.PRNGKey(0)

# dataset
D = random.bernoulli(key, 0.5, shape=[60000, 100])
# queries are the column indices we want to multiply
queries = random.permutation(key, np.array([comb for comb in itertools.combinations(np.arange(100), 3)]))[:1000]

# a single query: multiply the columns we care about and see what proportion of them are ones
def _single_query(D, query):
    return np.sum(np.prod(D[:, query], axis=1))/D.shape[0]

# generate a function that can compute result of some pre-determined subset of queries on the dataset, vmap over queries
def auto_batched_preserve_subset_statistic(queries):
    @jit
    def compute_statistic(D):
        return jit(vmap(_single_query, (None, 0)))(D, queries)
    return compute_statistic

# generate a function that can compute result of some pre-determined subset of queries on the dataset, hand vectorize
# over the queries
def hand_batched_preserve_subset_statistic(queries):
    @jit
    def compute_statistic(D):
        temp = np.array_split(queries, 10)
        return np.concatenate([
            np.prod(D[:, q], 2).sum(0) for q in temp
        ]) / D.shape[0]
    return compute_statistic

# hand batched statistic function
hb_compute_statistic = hand_batched_preserve_subset_statistic(queries)
# vmap/auto batched statistic function
ab_compute_statistic = auto_batched_preserve_subset_statistic(queries)
```

And to find how long it took:
```python
>>> %timeit ab_compute_statistic(D).block_until_ready()
2.39 ms ± 5.25 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

Versus...
```python
>>> %timeit hb_compute_statistic(D).block_until_ready()
660 µs ± 317 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

If I instead increased queries by a factor of 10, instead of the autobatching getting closer, both slowed down by the same factor (`hb_compute_statistic` -> 6 ms and `ab_compute_statistic` -> 23 ms).

I'm not entirely sure what's going wrong here (if anything is). I stuck to making sure I `jit` on `vmap` and not the other way.
_Originally posted by @AnkitSiva in https://github.com/google/jax/issues/6312#issuecomment-887686199_

**Comment by zhangqiaorjc** [comment]
@mattjj this is a followup from the original issue

**Comment by AnkitSiva** [comment]
@zhangqiaorjc could you add the `performance` tag to this issue? Thanks!

**Comment by ysngshn** [comment]
Any update on this? I also encountered a similar performance issue where vmap was noticeably slower than lax.scan (which would mean that doing batch processing is actually slower than doing a for loop to process one by one ...)

**Comment by jakevdp** [comment]
Hi @ysngshn - it looks like this has fallen a bit off the radar.

I tried to reproduce the original issue, and I do not see the speedup reported for the hand-batched case, so perhaps this issue has been fixed by a JAX or XLA update.

If you have another example of unexpected slowness in `vmap`, please post a minimal version and we can investigate - thanks!

**Comment by AnkitSiva** [comment]
Hey @jakevdp, thanks for getting back on this. Is the original issue you refer to here the code snippet I posted?

**Comment by jakevdp** [comment]
Yes, I find that `hb_compute_statistic` is slightly slower than `ab_compute_statistic` using JAX v0.2.19 and jaxlib v0.1.70.

Also, I don't think this is really an apples-to-apples comparison, because breaking your inputs into 10 separate batches is not the same as vmapping the full set of batches – perhaps you might expect the compiler to look at the queries, notice that the operation has no inter-dependence between rows, and use some heuristic to decide on an optimal batching of the inputs, but I'm not certain that XLA has that sort of optimization built-in to its compiler.

**Comment by ysngshn** [comment]
I tinkered around the code a bit, turns out that removing the deterministic flags solves the issue, however the result is no longer deterministic ... Also the batchsize for vmap was very small (2) so probably this is not the use case where vmap shines?

**Comment by jakevdp** [comment]
Can you give an example of what you refer to as "the batch size for vmap"? I don't know of any batching-related arguments that are supported by `vmap`. `vmap` is not really about batching, it's about vectorization.

**Comment by ysngshn** [comment]
Sorry for the "batchsize" term which might be confusing, it basically corresponds to the size of dimension specified by the `in_axes` argument of `vmap` call. In my use case I have multiple copies of network parameters and want to do the forward pass with all of them (preferably in parallel). The copies of network parameters are stored in a common PyTree, where dimension 0 of all its tree leaves are the in_axes. Here I notice that (I am testing with 2 copies of network parameters):

- if I opt for deterministic run, achieving this via `vmap` is noticeably slower than `lax.scan`
- if I remove the deterministic flags, `vmap ` is similarly fast compared to `lax.scan` (not 2 times faster as I would naively believe)

Is this actually the expected behavior?

**Comment by jakevdp** [comment]
Can you elaborate on why you expect vmap to be twice as fast as scan?

**Comment by ysngshn** [comment]
I am quite new to Jax so please correct me if I am wrong: from what I understand `vmap` is doing the 2 forward passes in parallel while `lax.scan` is basically a for loop and does them sequentially. Or is the compiler also making `lax.scan` run in parallel like doing parallel for? Then why would there be a huge performance penalty in the deterministic case only for `vmap` while `lax.scan` is more-or-less unaffected?

**Comment by jakevdp** [comment]
Thanks - I think you have a mistaken understanding of `vmap`. `vmap` does not relate to parallelization (you might be thinking of [`pmap`](https://jax.readthedocs.io/en/latest/jax.html#jax.pmap)) What `vmap` does is automatic vectorization. So, for example, if you have a program that you've written that processes a single row of a matrix, `vmap` will transform that to a vectorized operation across all rows of the matrix, without you having to rewrite your code.

A quick example; say you have this function that normalizes a vector:
```python
from jax import vmap
import jax.numpy as jnp

def process_row(x):
  return x / x.sum()

M = jnp.ones((10, 20))
```
If you want to apply this to all rows of a matrix in a vectorized fashion, you might write something like this:
```python
def process_all_rows(M):
  return M / M.sum(1).reshape(M.shape[0], 1)
```
We can use the `make_jaxpr` function to see the representation of the operations this function encodes:
```python
from jax import make_jaxpr

print(make_jaxpr(process_all_rows)(M))
# { lambda  ; a.
#   let b = reduce_sum[ axes=(1,) ] a
#       c = reshape[ dimensions=None
#                    new_sizes=(10, 1) ] b
#       d = div a c
#   in (d,) }
```
Now, as these row-wise functions get more complicated, doing this sort of manual vectorization can be complicated and error prone. Fortunately, the `vmap` transform exists and can do this for you!
```python
process_all_rows = vmap(process_row)

print(make_jaxpr(vmap(process_row))(M))
# { lambda  ; a.
#   let b = reduce_sum[ axes=(1,) ] a
#       c = reshape[ dimensions=None
#                    new_sizes=(10, 1) ] b
#       d = div a c
#   in (d,) }
```
You see that the `vmap`-transformed version results in exactly the same set of operations as the manually-vectorized version. This is the purpose of `vmap` – it's not about parallelization or batching of operations, but rather about automated vectorization.

**Comment by ysngshn** [comment]
Thanks a lot for the detailed explanation! And I apologize again for my poor choice of wording: I understand that `vmap` is actually "vectorizing" the operations (and I am using a single GPU so SIMD parallelization via `pmap` is probably not applicable anyway). Still, I would expected that the "vectorized" version by `vmap` would run faster on GPU than the sequential `lax.scan` version, in the spirit of "use vectorization instead of for loop" for tensor operations.

BTW, when using `lax.scan` with `carry=None`, will this be compiled into "parallel for" or is it always sequential for loop?

**Comment by jakevdp** [comment]
When you use `scan`, it is lowered to [`XLA:While`](https://www.tensorflow.org/xla/operation_semantics#while). I'm not sure how this is actually implemented on CPU or GPU and whether it takes advantage of any sort of parallelism.

**Comment by blacksph3re** [comment]
I have another code which is a little bit constructed but might help troubleshooting. On colab CPU, vmap is 4 times slower than a manual for loop, on my personal computer I get factor 10:
```
import jax
import jax.numpy as jnp
from jax.typing import ArrayLike

import time

def trapdoor(_i: int, val: ArrayLike) -> jax.Array:
    return (val ** 3) % 9973

def expensive_computation(x: ArrayLike) -> jax.Array:
    return jnp.sum(jax.lax.fori_loop(1, int(1e6), trapdoor, x % (jnp.arange(1024) + 2)))

expensive_computation_jitted = jax.jit(expensive_computation)
expensive_computation_vmapped = jax.vmap(expensive_computation, (0))

# Warmup so we don't measure compilation times
expensive_computation_vmapped(jnp.arange(5)).block_until_ready()

start = time.time()
res1 = expensive_computation_vmapped(jnp.arange(5)).block_until_ready()
print(time.time() - start)

# Warmup so we don't measure compilation times
expensive_computation_jitted(123)

start = time.time()
res2 = jnp.stack([expensive_computation_jitted(i) for i in range(5)]).block_until_ready()
print(time.time() - start)

assert all(res1 == res2)
```

On a M2 macbook, this prints out 
```
11.061187982559204
1.2404420375823975
```

On colab CPU, this prints
```
12.033756256103516
3.6273393630981445
```
However, on colab GPU (T4) it works as expected:
```
6.1040568351745605
27.996851444244385
```

**Comment by mattjj** [comment]
@blacksph3re I just ran your script on an AMD machine and got this:

```
$ python 7449.py
1.215203046798706
1.2163002490997314
```

It's very helpful to flag programs which have surprising performance! So thanks to all for chiming in here. However, I think these particular examples may be stale, and we should close this issue thread (but please open new ones as they arise).
